tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,253 @@
+///
+/// This registry is implemented using a [lock-free sharded slab][1], and is
+/// highly optimized for concurrent access.
+///
+/// [1]: https://docs.rs/crate/sharded-slab/0.0.5
+#[derive(Debug)]
+pub struct Registry {
+    spans: Slab<DataInner>,
+}
+
+/// Span data stored in a [`Registry`].
+///
+/// This definition  is intentionally kept small—only data pertaining to span
+/// relationships, span metadata, and active references are stored. Additional
+/// data, such as formatted fields, may be stored in the [extensions] typemap.
+///
+/// [`Registry`]: ../struct.Registry.html
+/// [extensions]: ../extensions/index.html
+#[derive(Debug)]
+pub struct Data<'a> {
+    inner: Guard<'a, DataInner>,
+}
+
+#[derive(Debug)]
+struct DataInner {
+    metadata: &'static Metadata<'static>,
+    parent: Option<Id>,
+    ref_count: AtomicUsize,
+    pub(crate) extensions: RwLock<ExtensionsInner>,
+}
+
+// === impl Registry ===
+
+impl Default for Registry {
+    fn default() -> Self {
+        Self { spans: Slab::new() }
+    }
+}
+
+#[inline]
+fn idx_to_id(idx: usize) -> Id {
+    Id::from_u64(idx as u64 + 1)
+}
+
+#[inline]
+fn id_to_idx(id: &Id) -> usize {
+    id.into_u64() as usize - 1
+}
+
+impl Registry {
+    fn insert(&self, s: DataInner) -> Option<usize> {
+        self.spans.insert(s)
+    }
+
+    fn get(&self, id: &Id) -> Option<Guard<'_, DataInner>> {
+        self.spans.get(id_to_idx(id))
+    }
+}
+
+thread_local! {
+    static CURRENT_SPANS: RefCell<SpanStack> = RefCell::new(SpanStack::new());
+}
+
+impl Subscriber for Registry {
+    fn register_callsite(&self, _: &'static Metadata<'static>) -> Interest {
+        Interest::always()
+    }
+
+    fn enabled(&self, _: &Metadata<'_>) -> bool {
+        true
+    }
+
+    #[inline]
+    fn new_span(&self, attrs: &span::Attributes<'_>) -> span::Id {
+        let parent = if attrs.is_root() {
+            None
+        } else if attrs.is_contextual() {
+            self.current_span().id().map(|id| self.clone_span(id))
+        } else {
+            attrs.parent().map(|id| self.clone_span(id))
+        };
+
+        let s = DataInner {
+            metadata: attrs.metadata(),
+            parent,
+            ref_count: AtomicUsize::new(1),
+            extensions: RwLock::new(ExtensionsInner::new()),
+        };
+        let id = self.insert(s).expect("Unable to allocate another span");
+        idx_to_id(id)
+    }
+
+    /// This is intentionally not implemented, as recording fields
+    /// on a span is the responsibility of layers atop of this registry.
+    #[inline]
+    fn record(&self, _: &span::Id, _: &span::Record<'_>) {}
+
+    fn record_follows_from(&self, _span: &span::Id, _follows: &span::Id) {}
+
+    /// This is intentionally not implemented, as recording events
+    /// is the responsibility of layers atop of this registry.
+    fn event(&self, _: &Event<'_>) {}
+
+    fn enter(&self, id: &span::Id) {
+        CURRENT_SPANS.with(|spans| {
```

Comment:
It seems odd to me that a type called `Registry` keeps track of the current span.

Reply:
It's part of the responsibility of a `Subscriber` to track the current span. Also, `Registry` exposes this to child layers through `Context`.